src/BreakpointViewer.cpp

```diff
@@ -438,278 +477,364 @@ void BreakpointViewer::setBreakState()
 	runState = false;
 }
 
-void BreakpointViewer::sync()
+void BreakpointViewer::stretchTable(BreakpointData::Type type)
 {
-	auto sa = ScopedAssign(userMode, false);
-	keepUnusedItems();
-
-	int bpRow = bpTableWidget->rowCount();
-	int wpRow = wpTableWidget->rowCount();
-	int cnRow = cnTableWidget->rowCount();
+	auto stretch = [](auto* table) {
+		// stretching will not work without sorting
+		table->setSortingEnabled(true);
+		table->resizeColumnsToContents();
+	};
+
+	if (type == BreakpointData::ALL) {
+		for (auto* table : tables) {
+			stretch(table);
+		}
+	} else {
+		stretch(tables[type]);
+	}
+}
 
-	if (breakpoints->breakpointCount() != -1) {
-		for (int index = 0; index < breakpoints->breakpointCount(); ++index) {
-			const Breakpoint& bp = breakpoints->getBreakpoint(index);
+std::map<QString, BreakpointData>::iterator
+BreakpointViewer::scanBreakpointData(const Breakpoint& bp)
+{
+	BreakpointData::Type type = bp.type == Breakpoint::BREAKPOINT ? BreakpointData::BREAKPOINT
+	                         : (bp.type == Breakpoint::CONDITION  ? BreakpointData::CONDITION : BreakpointData::WATCHPOINT);
+	if (maps[type].empty()) return maps[type].end();
+
+	// assumed premise: a new bp will never repeat same name again
+	if (auto it = maps[type].find(bp.id); it != maps[type].end() && it->second.index == -1) {
+		// assumed premise 2: if name remains the same, bp is exactly the same.
+		return it;
+	}
 
-			switch (bp.type) {
-			case Breakpoint::BREAKPOINT: {
-				bpRow = createTableRow(BREAKPOINT);
-				setBreakpointStatus(BREAKPOINT, bpRow, true);
+	return maps[type].end();
+}
 
-				bool currentSlot = breakpoints->inCurrentSlot(bp);
-				QBrush color = palette().brush(runState | currentSlot ?
-					QPalette::ColorGroup::Normal : QPalette::ColorGroup::Disabled,
-					QPalette::WindowText);
+void BreakpointViewer::connectBreakpointID(const QString& id, BreakpointData& data)
+{
+	maps[data.type].insert({id, data});
```

Comment:
(detail) An alternative is this (this is closer to the Qt version you had before):
```maps[...].try_emplace(id, data);```

There are some technical and semantic differences between `insert()`, `emplace()` and `try_emplace()`. As a rule of thumb, usually, `try_emplace()` is best when you pass separate key and value. But it doesn't really matter for a map with simple key/value types like this.

(No need to change this).

Reply:
`try_emplace` does nothing if the key already exists, but reusing an old key may be a unknown synchronization problem. Then I guess it should prompt the user with that error dialog and leave the failed set_bp in a unchecked status. I will write a fix for this feature.